File: MergeSort.cpp
#include <vector>
#include <cstdio>

using namespace std;

template <typename baseT>
class MergeSort{

private:
	vector<baseT> tmpv;

public:
	MergeSort(){

	}

	void sizing(vector<baseT> &arr){
		tmpv.resize(arr.size());
	}

	template <typename T1,typename T2>
	void merge(T1 l, T2 r, vector<baseT> &arr){
		if(l >= r) return;
		T1 mid = (l+r)/2;
		merge(l,mid,arr);
		merge(mid+1,r,arr);
		mergesort(l,r,arr);
	}

	template <typename T1, typename T2>
	void mergesort(T1 L, T2 R, vector<baseT> &arr){
		T1 l,r,idx,mid;
		l = idx = L;
		mid = (L+R)/2;
		r = mid+1;

		while(l <= mid && r <= R){
			if(arr[l] < arr[r]) tmpv[idx++] = arr[l++];
			else tmpv[idx++] = arr[r++];
		}

		while(l <= mid) tmpv[idx++] = arr[l++];
		while(r <= R) tmpv[idx++] = arr[r++];

		for(T1 i=L;i<=R;i++) arr[i] = tmpv[i];
	}

	void tmpToArrCopy(vector<baseT> &arr){
		for(int i=0;i<arr.size();i++) arr[i] = tmpv[i];
	}
};

File: MergeSort/MergeSort.java
public class MergeSort {
	private int[] tmparr;
	
	public void init(int length) {
		if(tmparr == null || tmparr.length < length) tmparr = new int[length];
	}

	public void sort(int[] arr,int l,int r){
		init(arr.length);
		split(arr,l,r);
	}

	private void split(int[] arr,int l,int r){
		if(l+1 >= r) return;

		int mid = (l+r)/2;
		split(arr,l,mid);
		split(arr,mid,r);
		mergeSort(arr,l,r);
	}

	private void mergeSort(int[] arr,int l,int r){
		int tmpl = l;
		int mid = (l+r)/2;
		int tmpr = mid;
		int idx = l;
		while(tmpl<mid && tmpr<r){
			if(arr[tmpl] < arr[tmpr]) tmparr[idx++] = arr[tmpl++];
			else tmparr[idx++] = arr[tmpr++];
		}

		while(tmpl<mid) tmparr[idx++] = arr[tmpl++];
		while(tmpr<r) tmparr[idx++] = arr[tmpr++];

		for(int i=l;i<r;i++){
			arr[i] = tmparr[i];
		}
	}
}
